ratis-server/src/main/java/org/apache/ratis/server/leader/LogAppenderBase.java

```diff
@@ -128,7 +128,8 @@ public AppendEntriesRequestProto newAppendEntriesRequest(long callId, boolean he
     final TermIndex previous = getPrevious(follower.getNextIndex());
     final long snapshotIndex = follower.getSnapshotIndex();
     final long heartbeatRemainingMs = getHeartbeatRemainingTimeMs();
-    if (heartbeatRemainingMs <= 0L || heartbeat) {
+    if ((heartbeatRemainingMs <= 0L && follower.getLastHeartbeatSendTime().elapsedTimeMs()
+        > getServer().properties().minRpcTimeoutMs()/4) || heartbeat) {
```

Comment:
Before the change, it will return when (heartbeatRemainingMs <= 0L).

After the change, when (heartbeatRemainingMs <= 0L and follower.getLastHeartbeatSendTime() < min/4), it won't return and run the else-clause.  It seems a problem.

Reply:
If  (heartbeatRemainingMs <= 0L and follower.getLastHeartbeatSendTime() < min/4),  which means we have sent out the heartbeat and have not received the response,  then we can go on with sending message.  Is this we expected?

Reply:
You are right.  The new logic seems correct.  Also, we return null when the buffer.isEmpty().

Not sure where is the bug now.